### PR TITLE
check host ports on join

### PR DESCRIPTION
**Pull request by mattkasun**

### CODE REVIEW

File: logic/hosts.go [hunk 1]
```diff
@@ -313,3 +313,53 @@ func GetRelatedHosts(hostID string) []models.Host {
...
+	proxyPort := h.ProxyListenPort
+	count := 0
+	reset := false
+	hosts, err := GetAllHosts()
+	if err != nil {
+		return
+	}
+	for _, host := range hosts {
+		if host.EndpointIP.Equal(h.EndpointIP) {
+			if host.ListenPort == h.ListenPort || host.ProxyListenPort == h.ProxyListenPort {
```
Comment (0xdcarns) [inline logic/hosts.go:331]: we need to compare if the listenport == listenport || listenport == proxylistenport && proxyport == proxyport || proxyport == listenport, as if you had 100 machines on same NAT they could start to overlap on default ports

File: logic/hosts.go [hunk 2]
```diff
@@ -313,3 +313,53 @@ func GetRelatedHosts(hostID string) []models.Host {
...
+	if (err != nil && !database.IsEmptyRecord(err)) || (err == nil) {
+		return true
+	}
+	return false
+}
+
+func updateListenPorts(h *models.Host) {
+	h.ListenPort++
+	h.ProxyListenPort++
+	if h.ListenPort > 65535 || h.ProxyListenPort > 65535 {
```
Comment (0xdcarns) [inline logic/hosts.go:361]: 65535 is used twice - use constant, there is probably one for a uint16 max number or something

**Review by 0xdcarns — CHANGES_REQUESTED** [review]

**Review by 0xdcarns — APPROVED** [review]

**Review by afeiszli — APPROVED** [review]

File: logic/hosts.go [hunk 3]
```diff
@@ -313,3 +318,60 @@ func GetRelatedHosts(hostID string) []models.Host {
...
+			}
+		}
+	}
+	return p
+}
+
+// HostExists - checks if given host already exists
+func HostExists(h *models.Host) bool {
+	_, err := GetHost(h.ID.String())
+	if (err != nil && !database.IsEmptyRecord(err)) || (err == nil) {
```
Comment (0xdcarns) [inline logic/hosts.go:366]: could just return this statement.

File: logic/hosts.go [hunk 4]
```diff
@@ -313,3 +318,60 @@ func GetRelatedHosts(hostID string) []models.Host {
...
+	h.ProxyListenPort = checkPort(h, h.ProxyListenPort)
+	// rerun if ports were set to same value
+	if h.ListenPort == h.ProxyListenPort {
+		updatePort(&h.ProxyListenPort)
+		h.ProxyListenPort = checkPort(h, h.ProxyListenPort)
+	}
+}
+
+func checkPort(h *models.Host, p int) int {
+	currentPort := h.ListenPort
```
Comment (0xdcarns) [inline logic/hosts.go:336]: if you set the proxylistenport to p, it will default to listenport? is this purposeful?

**Review by 0xdcarns — CHANGES_REQUESTED** [review]

**Comment by theguy951357** [comment]
just waiting on changes to be resolved before testing this one

File: logic/hosts.go (hunk 4 above)
Comment (mattkasun) [reply logic/hosts.go]: good catch.... that's what happens when you try to reduce code copying by combining two similar functions into one. 
fixed.

**Review by theguy951357 — APPROVED** [review]
was tested with https://github.com/gravitl/netclient/pull/138 
pass

**Review by theguy951357 — APPROVED** [review]
pass

**Comment by 0xdcarns** [comment]
logic converted in #1957

**Pull request closed without merging**